src/public/orders/sales-order-transaction.ts

```diff
@@ -0,0 +1,65 @@
+import { Transaction } from "../common";
+import { SalesOrderStatus } from "./enums";
+import { SalesOrderCustomFieldMappingPOJO } from "./sales-order-time-range";
+
+/**
+ * Session data that may include auth data set by a data driven auth processes.
+ * May include arbitrary data set by the developer.
+ * Must be JSON serializable. Any method may update the session data,
+ * such as renewing a session token or updating a timestamp.
+ */
+export type SalesOrderSession<T extends object = object> = T & {
+  /**
+  * Auth object set by a data driven auth processes.
+  */
+  readonly auth?: {
+    /**
+     * Populated by the data driven auth processes for basic auth integrations.
+     */
+    readonly username?: string;
+
+    /**
+     * Populated by the data driven auth processes for basic auth integrations.
+     */
+    readonly password?: string;
+
+    /**
+     * Populated by the data driven auth processes for OAuth integrations.
+     */
+    readonly accessToken?: string;
+
+    /**
+     * Populated by the data driven auth processes for API Key integrations.
+     */
+    readonly apiKey?: string;
+  }
+}
+
+/**
+ * The ShpEngine Connect passes this object to every method call. It provides information about the
+ * transaction being performed, including authentication, metadata, etc.
+ */
+export interface SalesOrderTransaction<T extends object = object> extends Transaction<T> {
+  /**
+   * Session data that may include auth data set by a data driven auth processes.
+   * May include arbitrary data set by the developer.
+   * Must be JSON serializable. Any method may update the session data,
+   * such as renewing a session token or updating a timestamp.
+   */
+  session: SalesOrderSession<T>;
+
+  /**
+   * A mapping of custom user-defined statuses to ShipEngine Connect statuses.
+   *
+   * Each key is a user-defined status string, and the value is the corresponding
+   * ShipEngine Connect order status.
+   */
+  readonly statusMappings?: Readonly<{
+    [key: string]: SalesOrderStatus;
+  }>;
+
+  /**
+   * A mapping custom user-defined fields to RequestedFulfillmentExtensions
+   */
+  readonly fieldMappings?: Readonly<SalesOrderCustomFieldMappingPOJO>;
```

Comment:
Why are we putting these fields on the `Transaction` object when they're only relevant to the export orders call?

Reply:
I have the same question, in our Order Source API spec only the export call has these statuses: https://github.com/shipstation/integrations-ecommerce/blob/master/order_source.yaml#L1032

Reply:
Custom mappings are defined at the app/user level so it makes more sense to place them in the Transaction which exists at the app/user level. It's a more extensible approach especially if we need to add more methods and/or more customizations in the future.

Reply:
The mappings are defined at the app/user level, but in service of a specific need in a single method. There is no evidence that a more extensible approach is needed or that there would be friction if new methods also needed this data in the future.

They should be defined as close to their need as possible. I agree that putting them on an object named `SalesOrderTimeRange` feels wrong, but I think that is more an implication that it is an inappropriate name. It should be something like `GetSalesOrdersByDateRequest` or `GetSalesOrdersByDateParams` if it is intended as a parameter object.

Reply:
Right, but there will most likely be other customizations to other methods in this and other (new) app types. Let's set the pattern of putting the data where it should be. As you pointed out, it is defined at the app/user level. The fact that there is only one method that currently needs such customization is just because it is the first one.

I know it's more work, but we should pay down the debt now rather than later.

Reply:
Is the goal for the `Transaction` interface to represent the union of all possible seller/store configurations that could be relevant to an integration? It still seems cluttered to go down the path of putting data specific to a single function into this object that's shared between multiple functions.

Reply:
No, we have the `SalesOrderTransaction`. As more app types are added we would extend `Transaction`.

Reply:
Where are we drawing the line that something is a "customization" vs something being a part of the signature of request?
Could the dateRange filter on the `SalesOrderByDate` method belong on the Transaction too?

Reply:
Good question. The distinction is the scope of the data. dateRange isn't applicable to other methods in the app. But, custom mappings could be applicable to other methods that reference a status.

Think of Transaction as the application Context. It is passed to every method called and should have app-level data. This is much like how middlewares or [Go's Context](https://golang.org/pkg/context/) work.

Reply:
> The fact that there is only one method that currently needs such customization is just because it is the first one.

> I know it's more work, but we should pay down the debt now rather than later.

@anthonyshull 

Do we know of a Marketplace we will/want to integrate with in the future (or one we already have) that uses order status for anything other than export orders? If so, then this extra work could be worth it.

If not, then the extra work could be wasted cycles. Additionally, having things moved into the transaction, that are only consumed for 1 action, could end up being a source of confusion & noise for developers using SE connect.

Reply:
The idea is to establish a pattern because there absolutely will be other information that exists at the user/app level amongst the various app types. So, by placing it on the Transaction we make it obvious to future developers where they can find such data and where they should place such data in their future work.

The worst case is when the user/app data is on the Transaction for carrier apps and warehousing apps, but in xyz method for order apps, etc.

I don't agree that it will cause confusion for developers. If they don't need the mappings then they just don't grab them off of the Transaction.

Reply:
I've discussed this with @jeversmann-shipstation and @anthonyshull and have come to a shared understanding:

The Connect SDK is a layer on top of the Order Source HTTP REST API to expose a Javascript/Typescript interface. It makes some editorial decisions about how the javascript developer works with the backend HTTP API. For example, it created the concept of the `Transaction` object and a `SalesOrderTimeRangePOJO` object. Those are not concepts in the HTTP API, but those SDK objects can be instantiated by the order source runtime using data from the incoming HTTP request body.

Since the Connect SDK created the `Transaction` concept it makes sense that it decides what data from the HTTP request body is exposed via that `Transaction` object. 

In the "get sales orders" case, the Order Source API payload will include custom status mapping data (as well as time range criteria, auth credentals, etc). The order source runtime can be changed to populate the Transaction object with the available custom status mapping data, so that the App developer accesses that data in the way the Connect SDK authors intend.

However, in the case of "shipment notification" or "order acknowledgement", the Order Source API payload will _not_ include custom status mapping data. In that case, the order source runtime can populate the Transaction object with an empty mapping (or null, if the SDK prefers/allows).

We are not changing the HTTP payloads for any Order Source API. We _will_ change the order source runtime mapping of the payloads to the SDK objects.